### DISCUSSION TITLE
Recommended audio parameters for training HifiGan for `tts_models/ja/kokoro/tacotron2-DDC` ?

### DISCUSSION THREAD

**Discussion by kaiidams** [coqui-ai/TTS#677] [General Q&A] [answered]
I'm trying to train HifiGan for `tts_models/ja/kokoro/tacotron2-DDC` as the default vocoder WaveGrad is a bit slow on CPU.
I started finetuning from `vocoder_models/en/vctk/hifigan_v2`. Though it has different audio parameters than the TTS model,
I thought some of weights are useful.

The model that was trained for 355000 steps, can generate realistic audio, but with some artifacts. It seems the model is not improving much after around 200000.

This colab converts a true audio with `ap.melspectrogram()` and then converts back with `vocoder_model.inference()`.
https://colab.research.google.com/drive/1xB-rDm6NUrWNUGPEWaAgrb6PFcMfSOpM?usp=sharing
Audios are also attached in this discussion.
[samples.zip](https://github.com/coqui-ai/TTS/files/6873533/samples.zip)

The audio parameters I use is,
```
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:50.0
 | > mel_fmax:7600.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:True
 | > trim_db:60
 | > do_sound_norm:False
 | > stats_path:../data/scale_stats.npy
 | > base:10
 | > hop_length:256
 | > win_length:1024
```

It is different from `vocoder_models/en/vctk/hifigan_v2` in `mel_fmin/fmax`, `preemphasis`, and `stats_path`.

Do you have any advice to reduce the artifacts from generated audio?
If the audio parameters are suspicious, then I would train TTS with new parameters.
Anyway I think it is good if models can share same parameters.

**Comment by erogol** [comment]
Can you also post your model config here?

**Reply by kaiidams** [reply]
This is my config. Thanks.
``` json
{
    "model": "hifigan",
    "run_name": "kokoro_hifigan",
    "run_description": "hifigan mean-var scaling",
    "epochs": 10000,
    "batch_size": 16,
    "eval_batch_size": null,
    "mixed_precision": false,
    "run_eval": true,
    "test_delay_epochs": 10,
    "print_eval": false,
    "print_step": 25,
    "tb_plot_step": 100,
    "tb_model_param_stats": true,
    "save_step": 5000,
    "checkpoint": true,
    "keep_all_best": false,
    "keep_after": 10000,
    "num_loader_workers": 8,
    "num_eval_loader_workers": 0,
    "use_noise_augment": false,
    "output_path": "../model",
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "do_trim_silence": true,
        "trim_db": 60,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 50.0,
        "mel_fmax": 7600.0,
        "spec_gain": 1,
        "signal_norm": true,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path": "../data/kokoro_hifigan/scale_stats.npy"
    },
    "eval_split_size": 10,
    "data_path": "../data/kokoro-speech-v1_1-small/wavs",
    "feature_path": null,
    "seq_len": 8192,
    "pad_short": 2000,
    "conv_pad": 0,
    "use_cache": true,
    "wd": 0.0,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "weight_decay": 0.0
    },
    "use_stft_loss": false,
    "use_subband_stft_loss": false,
    "use_mse_gan_loss": true,
    "use_hinge_gan_loss": false,
    "use_feat_match_loss": true,
    "use_l1_spec_loss": true,
    "stft_loss_weight": 0.0,
    "subband_stft_loss_weight": 0.0,
    "mse_G_loss_weight": 1.0,
    "hinge_G_loss_weight": 0.0,
    "feat_match_loss_weight": 10.0,
    "l1_spec_loss_weight": 45.0,
    "stft_loss_params": {
        "n_ffts": [
            1024,
            2048,
            512
        ],
        "hop_lengths": [
            120,
            240,
            50
        ],
        "win_lengths": [
            600,
            1200,
            240
        ]
    },
    "l1_spec_loss_params": {
        "use_mel": true,
        "sample_rate": 22050,
        "n_fft": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "n_mels": 80,
        "mel_fmin": 50.0,
        "mel_fmax": 7600.0
    },
    "target_loss": "loss_0",
    "grad_clip": null,
    "lr_gen": 0.0002,
    "lr_disc": 0.0002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999,
        "last_epoch": -1
    },
    "use_pqmf": false,
    "diff_samples_for_G_and_D": false,
    "discriminator_model": "hifigan_discriminator",
    "generator_model": "hifigan_generator",
    "generator_model_params": {
        "upsample_factors": [
            8,
            8,
            2,
            2
        ],
        "upsample_kernel_sizes": [
            16,
            16,
            4,
            4
        ],
        "upsample_initial_channel": 128,
        "resblock_kernel_sizes": [
            3,
            7,
            11
        ],
        "resblock_dilation_sizes": [
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ]
        ],
        "resblock_type": "1"
    },
    "lr": 0.0001
}
```

**Reply by erogol** [reply]
can you try `feat_match_loss_weight = 200`

**Reply by kaiidams** [reply]
@erogol Thank you for the advice.

This one is from the step 505000 after changing feat_match_loss_weight. The audio sounds very similar to me. 
[inference_505000.zip](https://github.com/coqui-ai/TTS/files/6889939/inference_505000.zip)

This is the training metrics.
```
--> STEP: 547/567 -- GLOBAL_STEP: 509275
     | > G_l1_spec_loss: 0.33710 
     | > G_mse_fake_loss: 0.42852 
     | > G_feat_match_loss: 0.06137 
     | > G_gen_loss: 15.16969 
     | > G_adv_loss: 12.70330 
     | > loss_0: 27.87299 
     | > grad_norm_0: 0.00000 
     | > D_mse_gan_loss: 0.36645 
     | > D_mse_gan_real_loss: 0.02284 
     | > D_mse_gan_fake_loss: 0.01128 
     | > loss_1: 0.36645 
     | > grad_norm_1: 0.00000 
     | > current_lr_0: 1.584135544966385e-34
     | > current_lr_1: 1.584135544966385e-34
     | > step_time: 0.5647
     | > loader_time: 0.0016
```

and this is before changing the config.

```
   --> STEP: 20/567 -- GLOBAL_STEP: 208500
     | > G_l1_spec_loss: 0.31882 
     | > G_mse_fake_loss: 0.61522 
     | > G_feat_match_loss: 0.10732 
     | > G_gen_loss: 14.34687 
     | > G_adv_loss: 1.68843 
     | > loss_0: 16.03530 
     | > grad_norm_0: 0.00000 
     | > D_mse_gan_loss: 0.22796 
     | > D_mse_gan_real_loss: 0.00017 
     | > D_mse_gan_fake_loss: 0.00206 
     | > loss_1: 0.22796 
     | > grad_norm_1: 0.00000 
     | > current_lr_0: 1.8522310219067017e-60
     | > current_lr_1: 1.8522310219067017e-60
     | > step_time: 0.3201
     | > loader_time: 0.0006
```

In both cases, LR is very small. I'm not sure if this is expected.

**Reply by erogol** [reply]
LR is too small. anything smaller than 1e-8 does not really do something.  Change the scheduler parameters in the config and let it decay less aggressively. Or use a larger initial lr.

**Reply by kaiidams** [reply]
@erogol Thanks. After updating both `feat_match_loss_weight ` and `lr_scheduler_xxx_params.gamma`, resetting `steps`, generated clips are sounds better. These are model files. Can you include these in the future release and update the default vocoder for Japanese Tacotron?

```
CONFIG_URL = 'https://drive.google.com/uc?id=1fzKQQL8QlNxbkWgvXL8hXYNkomT6QQFO' #@param {type:"string"}
MODEL_URL = 'https://drive.google.com/uc?id=10M4lGn6zCg-kSci9YuKYOpfyFO-3O8dX' #@param {type:"string"}
SCALE_STATS_URL = 'https://drive.google.com/uc?id=1UhYZIk2GpTV1m-rJvc1W6I6iTQylUYYP' #@param {type:"string"}
```

**Reply by erogol** [reply]
Thanks. Sure !!

do you want to serve it from gdrive or I can host with the other released models? 

gdrive has a band-width limit so it might cause trouble for you and the users when too many people downloading.

**Reply by kaiidams** [reply]
Thanks! I prefer to keep them with the other released models. Google doesn't give us free large storage spaces.

**Reply by erogol** [reply]
Ok I'll get the models and make them available with the next release. Thanks again :)

**Reply by skol101** [reply]
@kaiidams could you share your updated config.json as well as tell how did you manage to continue training the model?

**Reply by skol101** [reply]
> LR is too small. anything smaller than 1e-8 does not really do something. Change the scheduler parameters in the config and let it decay less aggressively. Or use a larger initial lr.

Well, so default parameters should actually be updated in the repo, because 1e-8 is reached after just about 10k-12k steps. 
Which LR param should be updated as there 3:
"lr_gen": 0.0001,
"lr_disc": 0.0001,
...
 "lr": 0.0001

**Reply by kaiidams** [reply]
@skol101 In my case, I started from the pretrained English model. The learning rate of the model is already decayed, so the fine-tuning didn't do anything. LR of my configuration is not very useful if you want to try another training. Probably someone who tries to train another model from scratch can figure out better LR.

**Reply by skol101** [reply]
I see. I'm just observing that HiFiGan training doesn't progress beyond 5k steps: "loss_0" (a target loss function of hifigan) converged to around 19 and not moving anywhere. And I don't understand what I'm doing wrong.

![Screenshot from 2021-11-30 17-21-47](https://user-images.githubusercontent.com/53978091/144075605-753f9edc-a8b9-49dd-97b6-05a9a85f6fe9.png)

**Comment by kaiidams** [comment, marked as answer]
https://github.com/coqui-ai/TTS/discussions/677#discussioncomment-1047361 answers the question.
